### PR TITLE
feat: Add a job-agnostic scheduling engine to nexusd

### PR DESCRIPTION
**Pull request by aintnostressin**
- `Job` trait + `run_job` scheduler loop (overrun-skip, MAX_SLEEP chunking, shutdown-responsive) backed by a `JobSpec` registry.
- `job_cron` gate + `JobConfig` (`[jobs.<name>]`, cron-only).
- `nexusd jobs run <name>` / `jobs list` CLI; the default daemon also drives scheduled jobs alongside webapi + watcher.
- Unit tests for the scheduler and `job_cron` validation.
- TrustRank job from #984

### CODE REVIEW

**Comment by aintnostressin** [comment]
@greptile review

File: nexusd/src/jobs/lock.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
+use async_trait::async_trait;
+use nexus_common::db::{release_lock, try_acquire_lock, RedisResult};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+/// TTL on a job's run lock, and the crash backstop. Must exceed the longest
+/// expected run, or a run outliving its lease could overlap another.
+const LOCK_TTL_SECS: u64 = 3600;
```
[the remark this answers is not in the document]
Comment (aintnostressin) [reply nexusd/src/jobs/lock.rs]: For now it's good enough. We need to ensure jobs don''t exceed this time .This is just a clean up mechanism in case lock is not released.

File: nexusd/src/jobs/scheduler.rs [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+            tokio::select! {
+                biased;
+
+                // Dropping the future leaves the lock to its TTL, but we're exiting anyway.
+                _ = wait_for_shutdown(&mut shutdown_rx) => {
+                    info!(job = name, "Shutdown received during run; abandoning it");
+                    return;
+                }
+                () = self.run_locked(name, job) => {}
+            }
```
[the remark this answers is not in the document]
Comment (aintnostressin) [reply nexusd/src/jobs/scheduler.rs:265]: fixed: [be8c707](https://github.com/pubky/pubky-nexus/pull/995/commits/be8c70771ce9dd206d2eb76651d44db35e07c6df)

File: nexusd/src/jobs/scheduler.rs [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+                return;
+            }
+        }
+    }
+}
+
+/// Parses `cron_expr` and checks it has at least one upcoming fire time,
+/// returning the [`Schedule`] or an error. Used to fail fast when building a job.
+pub fn validate_cron(cron_expr: &str) -> Result<Schedule, CronParseError> {
+    let schedule = Schedule::from_str(cron_expr).map_err(|e| CronParseError {
```
[the remark this answers is not in the document]
Comment (aintnostressin) [reply nexusd/src/jobs/scheduler.rs:167]: This is false positive. I've added a unit test in [e034088](https://github.com/pubky/pubky-nexus/pull/995/commits/e0340889ae5e7c2b56fb4be23489a940a9644827)

**Comment by aintnostressin** [comment]
@greptile review again

**Review by SHAcollision — COMMENTED** [review]
Overall the scheduler looks good and well tested, nice work.

One thing about landing it: I think we should not merge this alone to main. Right now the registry is empty, only #984 give it a real job, and the lock issue below only really happen when a job run. So maybe better we keep #984 on top of feat/jobs and land both together, or just fold #984 into feat/jobs and make one PR. I don't like to leave the framework alone in main with nobody using it.

File: nexusd/src/jobs/lock.rs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+use nexus_common::db::{release_lock, try_acquire_lock, RedisError};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use super::error::{LockError, LockResult};
+
+/// TTL on a job's run lock, and the crash backstop. Must exceed the longest
+/// expected run, or a run outliving its lease could overlap another.
+const LOCK_TTL_SECS: u64 = 3600;
```
Comment (SHAcollision) [inline nexusd/src/jobs/lock.rs]: I think we have one real thing here. The "no overlap" only work until the 1h TTL. `run()` have no timeout and we don't renew the lease, so if a job run more than 1h it lose the lock and another process can start the same job at the same time. To happen you need 2 process AND a >1h run so is not so common, but we promise no overlap so better we fix.

Maybe we don't need the full lease renew/fencing. Simpler idea: add a timeout in the `select!` of `run_locked` so the job can not run longer than the lease (if pass, we abandon + log), and make lease+timeout one config value. Like this the rule is always true and no extra renew loop. We can add a test on the boundary. Only if some job really need >1h we do the renew thing later.

File: nexus-common/src/config/job.rs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+use serde::{Deserialize, Serialize};
+
+/// Scheduling config for a cron job, keyed under `[jobs.<name>]`.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
+#[serde(default, deny_unknown_fields)]
+pub struct JobConfig {
+    /// 7-field cron (`sec min hour dom month dow [year]`, seconds-first).
```
Comment (SHAcollision) [inline nexus-common/src/config/job.rs]: nit: comment say "7-field" but we accept 6 and 7. Maybe reword.

File: nexusd/src/jobs/lock.rs (hunk 4 above)
Comment (aintnostressin) [reply nexusd/src/jobs/lock.rs]: done [49ffe59](https://github.com/pubky/pubky-nexus/pull/995/commits/49ffe593c6fcfa6325e9410475ba15de747f4b55)

File: nexus-common/src/config/job.rs (hunk 5 above)
Comment (aintnostressin) [reply nexus-common/src/config/job.rs]: done: [62a2a32](https://github.com/pubky/pubky-nexus/pull/995/commits/62a2a322ef8383cfda1b6141f54cfde922d7a826)

File: nexusd/src/jobs/lock.rs [hunk 6]
```diff
@@ -188,7 +189,7 @@ impl LockMetrics {
         ];
         self.operations_counter.add(1, &tags);
         self.duration_histogram
-            .record(duration.as_secs_f64(), &tags);
+            .record(duration.as_secs_f64() * 1000.0, &tags);
```
Comment (tipogi) [inline nexusd/src/jobs/lock.rs]: there is a helper function in `common/src/db/graph/instrumented.rs`, 
```rust
fn ms(d: Duration) -> f64 {
    d.as_secs_f64() * 1000.0
}
```

Comment (aintnostressin) [reply nexusd/src/jobs/lock.rs]: done in [21d5c26](https://github.com/pubky/pubky-nexus/pull/995/commits/21d5c2613c75d50e53aa3d8ff50239bce6bd27cf)

**Review by tipogi — APPROVED** [review]
ready!

**Review by SHAcollision — APPROVED** [review]
LGTM 

Heads up `trust.recompute.max_iterations_reached` is a permanent no-op in production.
